torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+            "The wait nodes are too far away from the comm node {comm_node}."
+        )
+
+    # Identify all the outputs
+    outputs: Dict[fx.Node, Any] = {}
+    nodes = collections.deque(wait_nodes)
+    while nodes:
+        node = nodes.popleft()
+        for user in node.users:
+            if isinstance(user, fx.Node) and user.name.startswith(non_end_users_nodes):
+                nodes.append(user)
+                node_list.append(user)
+            else:
+                outputs[node] = None
+                break
+
+    # TODO: populate all the tensor metadata and remove the default.
+    tensor_meta = inputs[0].meta.get("tensor_meta", None)
+    return CommBlock(
+        shape=tensor_meta.shape if tensor_meta else None,
+        node_list=node_list,
+        wait_nodes=wait_nodes,
+        comm_node=comm_node,
+        inputs=inputs,
+        outputs=outputs,
+    )
+
+
+def _create_meta_val(
+    fake_tensor_mode: FakeTensorMode,
+    val: FakeTensor,
+) -> FakeTensor:
+    # TODO: fix the memory_format
+    return FakeTensor(
+        fake_tensor_mode,
+        torch.empty(
+            val.shape,
+            dtype=val.dtype,
+            device="meta",
+            requires_grad=val.requires_grad,
+        ),
+        val.device,
+    )
+
+
+def _create_meta_tensor_meta(
+    fake_tensor_mode: FakeTensorMode,
+    val: FakeTensor,
+) -> FakeTensor:
+    return TensorMetadata(
+        shape=val.shape,
+        dtype=val.dtype,
+        requires_grad=val.requires_grad,
+        stride=val.stride,
+        # TODO: fix these value
+        memory_format=None,
+        is_quantized=False,
+        qparams={},
+    )
+
+
+def _call_function(
+    gm: fx.GraphModule,
+    fake_tensor_mode: FakeTensorMode,
+    meta_val: Optional[FakeTensor],
+    function: Any,
+    *args: Any,
+    **kwargs: Any,
+) -> fx.Node:
+    node = gm.graph.call_function(function, args, kwargs)
+
+    if meta_val is None:
+        flat_args, spec = tree_flatten((args, kwargs))
+        new_flat_args = []
+        memory_format = None
+        for arg in flat_args:
+            if not isinstance(arg, fx.Node):
+                new_flat_args.append(arg)
+                continue
+            val = arg.meta["val"]
+            new_flat_args.append(_create_meta_val(fake_tensor_mode, val))
+
+        fake_args, fake_kwargs = tree_unflatten(new_flat_args, spec)
+        new_meta_val = function(*fake_args, **fake_kwargs)
+    else:
+        new_meta_val = meta_val
+    node.meta["val"] = new_meta_val
+    node.meta["tensor_meta"] = _create_meta_tensor_meta(fake_tensor_mode, new_meta_val)
+    return node
+
+
+def _scatter_wait_result(
+    gm: fx.GraphModule,
+    fused_comm_block: CommBlock,
+    comm_blocks: List[CommBlock],
+    node_indices: Dict[fx.Node, int],
+) -> None:
+    last_wait_node_idx = 0
+    for node in gm.graph.nodes:
+        if node == fused_comm_block.comm_node:
+            break
+        last_wait_node_idx = max(
```

Comment:
just for discussion, we rely on the topological order given by fx.Graph, but do we need worry about that order is not a preferred topological order for us? E.g., a different topo order might allow us to delay wait a bit more? Will there be additional passes to further optimize that?

Reply:
Yes, as we discussed offline, the graph which generated by shared weights will not be optimized well without changing the topological sort order. I have another implementation which move the compute nodes between comm_ops and wait instead of moving wait. I feel that one can solve this issue, but that implementation still needs to be polished.